test/unit/error.spec.ts

```diff
@@ -292,3 +286,101 @@ test("return from nested finally", () => {
     `;
     expect(util.transpileAndExecute(code)).toBe("finally AB");
 });
+
+test("throw and catch custom error object", () => {
+    util.testFunction`
+      try {
+          throw { x: "Hello error object!" };
+      } catch (error) {
+          return error.x;
+      }
+    `.expectToMatchJsResult();
+});
+
+test.each(["Error", "RangeError", "ReferenceError", "SyntaxError", "TypeError", "URIError"])(
+    "throw builtin Errors as classes",
+    errorType => {
+        util.testFunction`
+            try {
+                throw new ${errorType}("message")
+            } catch (error) {
+                if (error instanceof Error) {
+                    return \`\${error}\`;
+                } else {
+                    throw TypeError();
+                }
+            }
+        `
+            .expectNoExecutionError()
+            .expectToMatchJsResult();
+    }
+);
+
+test.each(["Error", "RangeError", "ReferenceError", "SyntaxError", "TypeError", "URIError"])(
+    "throw builtin Errors as functions",
+    errorType => {
+        util.testFunction`
+            try {
+                throw ${errorType}("message")
+            } catch (error) {
+                if (error instanceof Error) {
+                    return \`\${error}\`;
+                } else {
+                    throw TypeError();
+                }
+            }
+        `
+            .expectNoExecutionError()
+            .expectToMatchJsResult();
+    }
+);
+
+test("subclass Error", () => {
+    util.testFunction`
+        class MyError extends Error {
+            name: "MyError"
+        }
+        
+        try {
+            throw new MyError();
+        } catch (error) {
+            if (error instanceof Error) {
+                return \'\$error\';
+            } else {
+                throw TypeError();
+            }
+        }
+    `
+        .expectNoExecutionError()
+        .expectToMatchJsResult();
+});
+
+test.each([["Error", false], ["Error", true], ["RangeError", false], ["MyError", false]] as Array<[string, boolean]>)(
+    "get stack from error",
+    (errorType, sourceMapTraceback) => {
+        const stack = util.testFunction`
+        class MyError extends Error {
+            name: "MyError"
+        }
+
+        function innerFunctionThatThrows() { throw new ${errorType}(); }
+        function outerFunctionThatThrows() { innerFunctionThatThrows(); }
+
+        try {
+            outerFunctionThatThrows();
+        } catch (error) {
+            if (error instanceof Error) {
+                return error.stack;
+            } else {
+                throw TypeError();
+            }
+        }
+    `
+            .setOptions({ sourceMapTraceback, luaLibImport: tstl.LuaLibImportKind.Inline })
```

Comment:
Could we somehow assert the stack to see the traceback is correctly applied?

Reply:
I'm not sure, AFAICT we don't have any tests that check the sourcemap transformation. Do they even work with non-file input?

Reply:
https://github.com/TypeScriptToLua/TypeScriptToLua/blob/master/test/unit/printer/sourcemaps.spec.ts

Reply:
Yeah I noticed that, but it’s not obvious to me if the output of debug.traceback is actually checked anywhere.

Reply:
Seems like this is a gap in our tests actually, I'm fine with it if you add a TODO here to add that to this test once we figure out how to do it. I made an issue to track this: #726